{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "continental-auction",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mon Feb  8 13:52:30 GMT 2021\n",
      "/ifs/obds-training/jan21/aneesha/python_datascience\n",
      "obds_py3              *  /ifs/obds-training/jan21/aneesha/conda/obds_conda/envs/obds_py3\n"
     ]
    }
   ],
   "source": [
    "!date\n",
    "!pwd\n",
    "! conda env list | grep '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "brief-patch",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import numpy\n",
    "import numpy as np\n",
    "# import pandas\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "monetary-calibration",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 23 entries, 0 to 22\n",
      "Data columns (total 4 columns):\n",
      " #   Column          Non-Null Count  Dtype \n",
      "---  ------          --------------  ----- \n",
      " 0   ref_seq_name    23 non-null     object\n",
      " 1   seq_length      23 non-null     int64 \n",
      " 2   mapped_reads    23 non-null     int64 \n",
      " 3   unmapped_reads  23 non-null     int64 \n",
      "dtypes: int64(3), object(1)\n",
      "memory usage: 864.0+ bytes\n",
      "RangeIndex(start=0, stop=23, step=1) (23, 4)\n"
     ]
    }
   ],
   "source": [
    "#Read in a idxstats table you have generated for one of your samples on the course and save to variable called idxdf\n",
    "idxdf = pd.read_csv('/ifs/obds-training/jan21/exercises/pandas/ERR1755082.idxstats', sep='\\t', header=None)\n",
    "\n",
    "#Assign column names to table\n",
    "idxdf.columns = ['ref_seq_name', 'seq_length', 'mapped_reads','unmapped_reads']\n",
    "idxdf \n",
    "\n",
    "#Apply the describe() and info() methods to the idxdf object\n",
    "idxdf.info()\n",
    "idxdf.describe()\n",
    "\n",
    "#Look at the shape and index attributes of idxdf\n",
    "print(idxdf.index, idxdf.shape)\n",
    "\n",
    "#From idxdf select rows 6-12 and create a new variable called idxdf2\n",
    "idxdf2 = idxdf[6:13]\n",
    "idxdf2\n",
    "idxdf \n",
    "\n",
    "#From idxdf make a new dataframe chr_mapped consisting of columns 1&3 (chr & mapped).\n",
    "#Save to a new file called mapped_reads_per_chr.tsv - make sure the output is tab seperated\n",
    "chr_mapped = idxdf[['ref_seq_name','mapped_reads']]\n",
    "chr_mapped.to_csv('mapped_read.tsv',sep='\\t', header=True, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "after-truth",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ref_seq_name    mapped_reads\n",
      "chr1    7505663\n",
      "chr10   4880751\n",
      "chr11   7452789\n",
      "chr12   3104950\n",
      "chr13   3222176\n",
      "chr14   3975495\n",
      "chr15   3796740\n",
      "chr16   2481586\n",
      "chr17   5794590\n",
      "chr18   2167389\n",
      "chr19   3573941\n",
      "chr2    6596073\n",
      "chr3    4062539\n",
      "chr4    5264332\n",
      "chr5    5440642\n",
      "chr6    4683974\n",
      "chr7    7687960\n",
      "chr8    4345076\n",
      "chr9    5147071\n",
      "chrM    2421340\n",
      "chrX    3431992\n",
      "chrY    310467\n",
      "\u001b[K:\u001b[Kped_read.tsv \u001b[m\u001b[K\u0007"
     ]
    }
   ],
   "source": [
    "!less mapped_read.tsv\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "provincial-affair",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_seq_name</th>\n",
       "      <th>seq_length</th>\n",
       "      <th>mapped_reads</th>\n",
       "      <th>unmapped_reads</th>\n",
       "      <th>reads_per_base</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>195471971</td>\n",
       "      <td>7505663</td>\n",
       "      <td>323256</td>\n",
       "      <td>0.038398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>130694993</td>\n",
       "      <td>4880751</td>\n",
       "      <td>248716</td>\n",
       "      <td>0.037345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>122082543</td>\n",
       "      <td>7452789</td>\n",
       "      <td>385535</td>\n",
       "      <td>0.061047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr12</td>\n",
       "      <td>120129022</td>\n",
       "      <td>3104950</td>\n",
       "      <td>145924</td>\n",
       "      <td>0.025847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr13</td>\n",
       "      <td>120421639</td>\n",
       "      <td>3222176</td>\n",
       "      <td>150791</td>\n",
       "      <td>0.026757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr14</td>\n",
       "      <td>124902244</td>\n",
       "      <td>3975495</td>\n",
       "      <td>191228</td>\n",
       "      <td>0.031829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr15</td>\n",
       "      <td>104043685</td>\n",
       "      <td>3796740</td>\n",
       "      <td>194000</td>\n",
       "      <td>0.036492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr16</td>\n",
       "      <td>98207768</td>\n",
       "      <td>2481586</td>\n",
       "      <td>122178</td>\n",
       "      <td>0.025269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr17</td>\n",
       "      <td>94987271</td>\n",
       "      <td>5794590</td>\n",
       "      <td>280652</td>\n",
       "      <td>0.061004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr18</td>\n",
       "      <td>90702639</td>\n",
       "      <td>2167389</td>\n",
       "      <td>103901</td>\n",
       "      <td>0.023896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr19</td>\n",
       "      <td>61431566</td>\n",
       "      <td>3573941</td>\n",
       "      <td>183566</td>\n",
       "      <td>0.058178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr2</td>\n",
       "      <td>182113224</td>\n",
       "      <td>6596073</td>\n",
       "      <td>320953</td>\n",
       "      <td>0.036220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr3</td>\n",
       "      <td>160039680</td>\n",
       "      <td>4062539</td>\n",
       "      <td>197868</td>\n",
       "      <td>0.025385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr4</td>\n",
       "      <td>156508116</td>\n",
       "      <td>5264332</td>\n",
       "      <td>262146</td>\n",
       "      <td>0.033636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr5</td>\n",
       "      <td>151834684</td>\n",
       "      <td>5440642</td>\n",
       "      <td>263428</td>\n",
       "      <td>0.035833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr6</td>\n",
       "      <td>149736546</td>\n",
       "      <td>4683974</td>\n",
       "      <td>223154</td>\n",
       "      <td>0.031281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr7</td>\n",
       "      <td>145441459</td>\n",
       "      <td>7687960</td>\n",
       "      <td>361209</td>\n",
       "      <td>0.052859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr8</td>\n",
       "      <td>129401213</td>\n",
       "      <td>4345076</td>\n",
       "      <td>212528</td>\n",
       "      <td>0.033578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr9</td>\n",
       "      <td>124595110</td>\n",
       "      <td>5147071</td>\n",
       "      <td>247008</td>\n",
       "      <td>0.041310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chrM</td>\n",
       "      <td>16299</td>\n",
       "      <td>2421340</td>\n",
       "      <td>80238</td>\n",
       "      <td>148.557580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chrX</td>\n",
       "      <td>171031299</td>\n",
       "      <td>3431992</td>\n",
       "      <td>151970</td>\n",
       "      <td>0.020066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chrY</td>\n",
       "      <td>91744698</td>\n",
       "      <td>310467</td>\n",
       "      <td>15700</td>\n",
       "      <td>0.003384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2305942</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ref_seq_name  seq_length  mapped_reads  unmapped_reads  reads_per_base\n",
       "0          chr1   195471971       7505663          323256        0.038398\n",
       "1         chr10   130694993       4880751          248716        0.037345\n",
       "2         chr11   122082543       7452789          385535        0.061047\n",
       "3         chr12   120129022       3104950          145924        0.025847\n",
       "4         chr13   120421639       3222176          150791        0.026757\n",
       "5         chr14   124902244       3975495          191228        0.031829\n",
       "6         chr15   104043685       3796740          194000        0.036492\n",
       "7         chr16    98207768       2481586          122178        0.025269\n",
       "8         chr17    94987271       5794590          280652        0.061004\n",
       "9         chr18    90702639       2167389          103901        0.023896\n",
       "10        chr19    61431566       3573941          183566        0.058178\n",
       "11         chr2   182113224       6596073          320953        0.036220\n",
       "12         chr3   160039680       4062539          197868        0.025385\n",
       "13         chr4   156508116       5264332          262146        0.033636\n",
       "14         chr5   151834684       5440642          263428        0.035833\n",
       "15         chr6   149736546       4683974          223154        0.031281\n",
       "16         chr7   145441459       7687960          361209        0.052859\n",
       "17         chr8   129401213       4345076          212528        0.033578\n",
       "18         chr9   124595110       5147071          247008        0.041310\n",
       "19         chrM       16299       2421340           80238      148.557580\n",
       "20         chrX   171031299       3431992          151970        0.020066\n",
       "21         chrY    91744698        310467           15700        0.003384\n",
       "22            *           0             0         2305942             NaN"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Create a new column for idxdf called 'reads_per_base' by dividing number of mapped reads by length column\n",
    "idxdf = idxdf.assign(reads_per_base = idxdf['mapped_reads'] / idxdf['seq_length'])\n",
    "idxdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "protecting-syria",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_seq_name</th>\n",
       "      <th>seq_length</th>\n",
       "      <th>mapped_reads</th>\n",
       "      <th>unmapped_reads</th>\n",
       "      <th>reads_per_base</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>122082543</td>\n",
       "      <td>7452789</td>\n",
       "      <td>385535</td>\n",
       "      <td>0.061047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr17</td>\n",
       "      <td>94987271</td>\n",
       "      <td>5794590</td>\n",
       "      <td>280652</td>\n",
       "      <td>0.061004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr19</td>\n",
       "      <td>61431566</td>\n",
       "      <td>3573941</td>\n",
       "      <td>183566</td>\n",
       "      <td>0.058178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr7</td>\n",
       "      <td>145441459</td>\n",
       "      <td>7687960</td>\n",
       "      <td>361209</td>\n",
       "      <td>0.052859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chrM</td>\n",
       "      <td>16299</td>\n",
       "      <td>2421340</td>\n",
       "      <td>80238</td>\n",
       "      <td>148.557580</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ref_seq_name  seq_length  mapped_reads  unmapped_reads  reads_per_base\n",
       "2         chr11   122082543       7452789          385535        0.061047\n",
       "8         chr17    94987271       5794590          280652        0.061004\n",
       "10        chr19    61431566       3573941          183566        0.058178\n",
       "16         chr7   145441459       7687960          361209        0.052859\n",
       "19         chrM       16299       2421340           80238      148.557580"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Subset the rows that have reads_per_base values greater than 0.05\n",
    "subset_idxdf = idxdf[idxdf.reads_per_base > 0.05]\n",
    "subset_idxdf\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "wound-addition",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_seq_name</th>\n",
       "      <th>seq_length</th>\n",
       "      <th>mapped_reads</th>\n",
       "      <th>unmapped_reads</th>\n",
       "      <th>reads_per_base</th>\n",
       "      <th>length_Mbs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>195471971</td>\n",
       "      <td>7505663</td>\n",
       "      <td>323256</td>\n",
       "      <td>0.038398</td>\n",
       "      <td>195.471971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>130694993</td>\n",
       "      <td>4880751</td>\n",
       "      <td>248716</td>\n",
       "      <td>0.037345</td>\n",
       "      <td>130.694993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>122082543</td>\n",
       "      <td>7452789</td>\n",
       "      <td>385535</td>\n",
       "      <td>0.061047</td>\n",
       "      <td>122.082543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr12</td>\n",
       "      <td>120129022</td>\n",
       "      <td>3104950</td>\n",
       "      <td>145924</td>\n",
       "      <td>0.025847</td>\n",
       "      <td>120.129022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr13</td>\n",
       "      <td>120421639</td>\n",
       "      <td>3222176</td>\n",
       "      <td>150791</td>\n",
       "      <td>0.026757</td>\n",
       "      <td>120.421639</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr14</td>\n",
       "      <td>124902244</td>\n",
       "      <td>3975495</td>\n",
       "      <td>191228</td>\n",
       "      <td>0.031829</td>\n",
       "      <td>124.902244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr15</td>\n",
       "      <td>104043685</td>\n",
       "      <td>3796740</td>\n",
       "      <td>194000</td>\n",
       "      <td>0.036492</td>\n",
       "      <td>104.043685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr16</td>\n",
       "      <td>98207768</td>\n",
       "      <td>2481586</td>\n",
       "      <td>122178</td>\n",
       "      <td>0.025269</td>\n",
       "      <td>98.207768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr17</td>\n",
       "      <td>94987271</td>\n",
       "      <td>5794590</td>\n",
       "      <td>280652</td>\n",
       "      <td>0.061004</td>\n",
       "      <td>94.987271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr18</td>\n",
       "      <td>90702639</td>\n",
       "      <td>2167389</td>\n",
       "      <td>103901</td>\n",
       "      <td>0.023896</td>\n",
       "      <td>90.702639</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr19</td>\n",
       "      <td>61431566</td>\n",
       "      <td>3573941</td>\n",
       "      <td>183566</td>\n",
       "      <td>0.058178</td>\n",
       "      <td>61.431566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr2</td>\n",
       "      <td>182113224</td>\n",
       "      <td>6596073</td>\n",
       "      <td>320953</td>\n",
       "      <td>0.036220</td>\n",
       "      <td>182.113224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr3</td>\n",
       "      <td>160039680</td>\n",
       "      <td>4062539</td>\n",
       "      <td>197868</td>\n",
       "      <td>0.025385</td>\n",
       "      <td>160.039680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr4</td>\n",
       "      <td>156508116</td>\n",
       "      <td>5264332</td>\n",
       "      <td>262146</td>\n",
       "      <td>0.033636</td>\n",
       "      <td>156.508116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr5</td>\n",
       "      <td>151834684</td>\n",
       "      <td>5440642</td>\n",
       "      <td>263428</td>\n",
       "      <td>0.035833</td>\n",
       "      <td>151.834684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr6</td>\n",
       "      <td>149736546</td>\n",
       "      <td>4683974</td>\n",
       "      <td>223154</td>\n",
       "      <td>0.031281</td>\n",
       "      <td>149.736546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr7</td>\n",
       "      <td>145441459</td>\n",
       "      <td>7687960</td>\n",
       "      <td>361209</td>\n",
       "      <td>0.052859</td>\n",
       "      <td>145.441459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr8</td>\n",
       "      <td>129401213</td>\n",
       "      <td>4345076</td>\n",
       "      <td>212528</td>\n",
       "      <td>0.033578</td>\n",
       "      <td>129.401213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr9</td>\n",
       "      <td>124595110</td>\n",
       "      <td>5147071</td>\n",
       "      <td>247008</td>\n",
       "      <td>0.041310</td>\n",
       "      <td>124.595110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chrM</td>\n",
       "      <td>16299</td>\n",
       "      <td>2421340</td>\n",
       "      <td>80238</td>\n",
       "      <td>148.557580</td>\n",
       "      <td>0.016299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chrX</td>\n",
       "      <td>171031299</td>\n",
       "      <td>3431992</td>\n",
       "      <td>151970</td>\n",
       "      <td>0.020066</td>\n",
       "      <td>171.031299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chrY</td>\n",
       "      <td>91744698</td>\n",
       "      <td>310467</td>\n",
       "      <td>15700</td>\n",
       "      <td>0.003384</td>\n",
       "      <td>91.744698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2305942</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ref_seq_name  seq_length  mapped_reads  unmapped_reads  reads_per_base  \\\n",
       "0          chr1   195471971       7505663          323256        0.038398   \n",
       "1         chr10   130694993       4880751          248716        0.037345   \n",
       "2         chr11   122082543       7452789          385535        0.061047   \n",
       "3         chr12   120129022       3104950          145924        0.025847   \n",
       "4         chr13   120421639       3222176          150791        0.026757   \n",
       "5         chr14   124902244       3975495          191228        0.031829   \n",
       "6         chr15   104043685       3796740          194000        0.036492   \n",
       "7         chr16    98207768       2481586          122178        0.025269   \n",
       "8         chr17    94987271       5794590          280652        0.061004   \n",
       "9         chr18    90702639       2167389          103901        0.023896   \n",
       "10        chr19    61431566       3573941          183566        0.058178   \n",
       "11         chr2   182113224       6596073          320953        0.036220   \n",
       "12         chr3   160039680       4062539          197868        0.025385   \n",
       "13         chr4   156508116       5264332          262146        0.033636   \n",
       "14         chr5   151834684       5440642          263428        0.035833   \n",
       "15         chr6   149736546       4683974          223154        0.031281   \n",
       "16         chr7   145441459       7687960          361209        0.052859   \n",
       "17         chr8   129401213       4345076          212528        0.033578   \n",
       "18         chr9   124595110       5147071          247008        0.041310   \n",
       "19         chrM       16299       2421340           80238      148.557580   \n",
       "20         chrX   171031299       3431992          151970        0.020066   \n",
       "21         chrY    91744698        310467           15700        0.003384   \n",
       "22            *           0             0         2305942             NaN   \n",
       "\n",
       "    length_Mbs  \n",
       "0   195.471971  \n",
       "1   130.694993  \n",
       "2   122.082543  \n",
       "3   120.129022  \n",
       "4   120.421639  \n",
       "5   124.902244  \n",
       "6   104.043685  \n",
       "7    98.207768  \n",
       "8    94.987271  \n",
       "9    90.702639  \n",
       "10   61.431566  \n",
       "11  182.113224  \n",
       "12  160.039680  \n",
       "13  156.508116  \n",
       "14  151.834684  \n",
       "15  149.736546  \n",
       "16  145.441459  \n",
       "17  129.401213  \n",
       "18  124.595110  \n",
       "19    0.016299  \n",
       "20  171.031299  \n",
       "21   91.744698  \n",
       "22    0.000000  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Create a new column of length / 1 million to give a length_Mbs column (length in Megabases)\n",
    "from IPython.display import display, HTML\n",
    "idxdf['length_Mbs']= idxdf['seq_length']/1000000\n",
    "display(idxdf) \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "infinite-savannah",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_seq_name</th>\n",
       "      <th>seq_length</th>\n",
       "      <th>mapped_reads</th>\n",
       "      <th>unmapped_reads</th>\n",
       "      <th>length_Mbs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>195471971</td>\n",
       "      <td>7505663</td>\n",
       "      <td>323256</td>\n",
       "      <td>195.471971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>130694993</td>\n",
       "      <td>4880751</td>\n",
       "      <td>248716</td>\n",
       "      <td>130.694993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>122082543</td>\n",
       "      <td>7452789</td>\n",
       "      <td>385535</td>\n",
       "      <td>122.082543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr12</td>\n",
       "      <td>120129022</td>\n",
       "      <td>3104950</td>\n",
       "      <td>145924</td>\n",
       "      <td>120.129022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr13</td>\n",
       "      <td>120421639</td>\n",
       "      <td>3222176</td>\n",
       "      <td>150791</td>\n",
       "      <td>120.421639</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr14</td>\n",
       "      <td>124902244</td>\n",
       "      <td>3975495</td>\n",
       "      <td>191228</td>\n",
       "      <td>124.902244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr15</td>\n",
       "      <td>104043685</td>\n",
       "      <td>3796740</td>\n",
       "      <td>194000</td>\n",
       "      <td>104.043685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr16</td>\n",
       "      <td>98207768</td>\n",
       "      <td>2481586</td>\n",
       "      <td>122178</td>\n",
       "      <td>98.207768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr17</td>\n",
       "      <td>94987271</td>\n",
       "      <td>5794590</td>\n",
       "      <td>280652</td>\n",
       "      <td>94.987271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr18</td>\n",
       "      <td>90702639</td>\n",
       "      <td>2167389</td>\n",
       "      <td>103901</td>\n",
       "      <td>90.702639</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr19</td>\n",
       "      <td>61431566</td>\n",
       "      <td>3573941</td>\n",
       "      <td>183566</td>\n",
       "      <td>61.431566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr2</td>\n",
       "      <td>182113224</td>\n",
       "      <td>6596073</td>\n",
       "      <td>320953</td>\n",
       "      <td>182.113224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr3</td>\n",
       "      <td>160039680</td>\n",
       "      <td>4062539</td>\n",
       "      <td>197868</td>\n",
       "      <td>160.039680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr4</td>\n",
       "      <td>156508116</td>\n",
       "      <td>5264332</td>\n",
       "      <td>262146</td>\n",
       "      <td>156.508116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr5</td>\n",
       "      <td>151834684</td>\n",
       "      <td>5440642</td>\n",
       "      <td>263428</td>\n",
       "      <td>151.834684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr6</td>\n",
       "      <td>149736546</td>\n",
       "      <td>4683974</td>\n",
       "      <td>223154</td>\n",
       "      <td>149.736546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr7</td>\n",
       "      <td>145441459</td>\n",
       "      <td>7687960</td>\n",
       "      <td>361209</td>\n",
       "      <td>145.441459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr8</td>\n",
       "      <td>129401213</td>\n",
       "      <td>4345076</td>\n",
       "      <td>212528</td>\n",
       "      <td>129.401213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr9</td>\n",
       "      <td>124595110</td>\n",
       "      <td>5147071</td>\n",
       "      <td>247008</td>\n",
       "      <td>124.595110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chrM</td>\n",
       "      <td>16299</td>\n",
       "      <td>2421340</td>\n",
       "      <td>80238</td>\n",
       "      <td>0.016299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chrX</td>\n",
       "      <td>171031299</td>\n",
       "      <td>3431992</td>\n",
       "      <td>151970</td>\n",
       "      <td>171.031299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chrY</td>\n",
       "      <td>91744698</td>\n",
       "      <td>310467</td>\n",
       "      <td>15700</td>\n",
       "      <td>91.744698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2305942</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ref_seq_name  seq_length  mapped_reads  unmapped_reads  length_Mbs\n",
       "0          chr1   195471971       7505663          323256  195.471971\n",
       "1         chr10   130694993       4880751          248716  130.694993\n",
       "2         chr11   122082543       7452789          385535  122.082543\n",
       "3         chr12   120129022       3104950          145924  120.129022\n",
       "4         chr13   120421639       3222176          150791  120.421639\n",
       "5         chr14   124902244       3975495          191228  124.902244\n",
       "6         chr15   104043685       3796740          194000  104.043685\n",
       "7         chr16    98207768       2481586          122178   98.207768\n",
       "8         chr17    94987271       5794590          280652   94.987271\n",
       "9         chr18    90702639       2167389          103901   90.702639\n",
       "10        chr19    61431566       3573941          183566   61.431566\n",
       "11         chr2   182113224       6596073          320953  182.113224\n",
       "12         chr3   160039680       4062539          197868  160.039680\n",
       "13         chr4   156508116       5264332          262146  156.508116\n",
       "14         chr5   151834684       5440642          263428  151.834684\n",
       "15         chr6   149736546       4683974          223154  149.736546\n",
       "16         chr7   145441459       7687960          361209  145.441459\n",
       "17         chr8   129401213       4345076          212528  129.401213\n",
       "18         chr9   124595110       5147071          247008  124.595110\n",
       "19         chrM       16299       2421340           80238    0.016299\n",
       "20         chrX   171031299       3431992          151970  171.031299\n",
       "21         chrY    91744698        310467           15700   91.744698\n",
       "22            *           0             0         2305942    0.000000"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Subset the rows that have reads_per_base values greater than 0.05 and a chromosome length under 100 Mb\n",
    "idxdf[(idxdf['reads_per_base'] > 0.05) & (idxdf['length_Mbs'] < 100)]\n",
    "\n",
    "#Delete the reads_per_base column\n",
    "idxdf = idxdf.drop('reads_per_base', axis=1)\n",
    "\n",
    "idxdf\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "greek-spencer",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/ifs/obds-training/jan21/aneesha/conda/obds_conda/envs/obds_py3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3263: DtypeWarning: Columns (0) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  if (await self.run_code(code, result,  async_=asy)):\n"
     ]
    }
   ],
   "source": [
    "#read in ensembl gtf files to create gtf\n",
    "gtf = pd.read_table('/ifs/obds-training/jan21/exercises/pandas/Mus_musculus.GRCm38.102.chr.gtf.gz', sep='\\t', header=None, comment='#')\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "pretty-tribe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        seq_name   source     feature    start      end score strand frame  \\\n",
      "0              1   havana        gene  3073253  3074322     .      +     .   \n",
      "1              1   havana  transcript  3073253  3074322     .      +     .   \n",
      "2              1   havana        exon  3073253  3074322     .      +     .   \n",
      "3              1  ensembl        gene  3102016  3102125     .      +     .   \n",
      "4              1  ensembl  transcript  3102016  3102125     .      +     .   \n",
      "...          ...      ...         ...      ...      ...   ...    ...   ...   \n",
      "1868658       MT    insdc  transcript    15289    15355     .      +     .   \n",
      "1868659       MT    insdc        exon    15289    15355     .      +     .   \n",
      "1868660       MT    insdc        gene    15356    15422     .      -     .   \n",
      "1868661       MT    insdc  transcript    15356    15422     .      -     .   \n",
      "1868662       MT    insdc        exon    15356    15422     .      -     .   \n",
      "\n",
      "                                                 attribute  \n",
      "0        gene_id \"ENSMUSG00000102693\"; gene_version \"1\"...  \n",
      "1        gene_id \"ENSMUSG00000102693\"; gene_version \"1\"...  \n",
      "2        gene_id \"ENSMUSG00000102693\"; gene_version \"1\"...  \n",
      "3        gene_id \"ENSMUSG00000064842\"; gene_version \"1\"...  \n",
      "4        gene_id \"ENSMUSG00000064842\"; gene_version \"1\"...  \n",
      "...                                                    ...  \n",
      "1868658  gene_id \"ENSMUSG00000064371\"; gene_version \"1\"...  \n",
      "1868659  gene_id \"ENSMUSG00000064371\"; gene_version \"1\"...  \n",
      "1868660  gene_id \"ENSMUSG00000064372\"; gene_version \"1\"...  \n",
      "1868661  gene_id \"ENSMUSG00000064372\"; gene_version \"1\"...  \n",
      "1868662  gene_id \"ENSMUSG00000064372\"; gene_version \"1\"...  \n",
      "\n",
      "[1868663 rows x 9 columns]\n",
      "['havana' 'ensembl' 'ensembl_havana' 'mirbase' 'insdc']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_name</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3205901</td>\n",
       "      <td>3671498</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000051951\"; gene_version \"5\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3999557</td>\n",
       "      <td>4409241</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000025900\"; gene_version \"13...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>4490931</td>\n",
       "      <td>4497354</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000025902\"; gene_version \"13...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>4773206</td>\n",
       "      <td>4785739</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000033845\"; gene_version \"13...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>4807788</td>\n",
       "      <td>4848410</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000025903\"; gene_version \"14...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868189</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>60864051</td>\n",
       "      <td>60874556</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000024997\"; gene_version \"7\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868209</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>60889749</td>\n",
       "      <td>61095968</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000003228\"; gene_version \"9\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868272</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>61053840</td>\n",
       "      <td>61140840</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000074733\"; gene_version \"14...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868395</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>61174686</td>\n",
       "      <td>61176309</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000094649\"; gene_version \"1\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868407</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>61223957</td>\n",
       "      <td>61228429</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000059326\"; gene_version \"7\"...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22436 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seq_name          source feature     start       end score strand  \\\n",
       "6              1  ensembl_havana    gene   3205901   3671498     .      -   \n",
       "70             1  ensembl_havana    gene   3999557   4409241     .      -   \n",
       "160            1  ensembl_havana    gene   4490931   4497354     .      -   \n",
       "283            1  ensembl_havana    gene   4773206   4785739     .      -   \n",
       "354            1  ensembl_havana    gene   4807788   4848410     .      +   \n",
       "...          ...             ...     ...       ...       ...   ...    ...   \n",
       "1868189       19  ensembl_havana    gene  60864051  60874556     .      -   \n",
       "1868209       19  ensembl_havana    gene  60889749  61095968     .      +   \n",
       "1868272       19  ensembl_havana    gene  61053840  61140840     .      -   \n",
       "1868395       19  ensembl_havana    gene  61174686  61176309     .      -   \n",
       "1868407       19  ensembl_havana    gene  61223957  61228429     .      -   \n",
       "\n",
       "        frame                                          attribute  \n",
       "6           .  gene_id \"ENSMUSG00000051951\"; gene_version \"5\"...  \n",
       "70          .  gene_id \"ENSMUSG00000025900\"; gene_version \"13...  \n",
       "160         .  gene_id \"ENSMUSG00000025902\"; gene_version \"13...  \n",
       "283         .  gene_id \"ENSMUSG00000033845\"; gene_version \"13...  \n",
       "354         .  gene_id \"ENSMUSG00000025903\"; gene_version \"14...  \n",
       "...       ...                                                ...  \n",
       "1868189     .  gene_id \"ENSMUSG00000024997\"; gene_version \"7\"...  \n",
       "1868209     .  gene_id \"ENSMUSG00000003228\"; gene_version \"9\"...  \n",
       "1868272     .  gene_id \"ENSMUSG00000074733\"; gene_version \"14...  \n",
       "1868395     .  gene_id \"ENSMUSG00000094649\"; gene_version \"1\"...  \n",
       "1868407     .  gene_id \"ENSMUSG00000059326\"; gene_version \"7\"...  \n",
       "\n",
       "[22436 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_name</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3205901</td>\n",
       "      <td>3671498</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000051951\"; gene_version \"5\"...</td>\n",
       "      <td>465597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3999557</td>\n",
       "      <td>4409241</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000025900\"; gene_version \"13...</td>\n",
       "      <td>409684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>4490931</td>\n",
       "      <td>4497354</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000025902\"; gene_version \"13...</td>\n",
       "      <td>6423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>4773206</td>\n",
       "      <td>4785739</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000033845\"; gene_version \"13...</td>\n",
       "      <td>12533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>4807788</td>\n",
       "      <td>4848410</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000025903\"; gene_version \"14...</td>\n",
       "      <td>40622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868189</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>60864051</td>\n",
       "      <td>60874556</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000024997\"; gene_version \"7\"...</td>\n",
       "      <td>10505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868209</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>60889749</td>\n",
       "      <td>61095968</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000003228\"; gene_version \"9\"...</td>\n",
       "      <td>206219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868272</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>61053840</td>\n",
       "      <td>61140840</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000074733\"; gene_version \"14...</td>\n",
       "      <td>87000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868395</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>61174686</td>\n",
       "      <td>61176309</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000094649\"; gene_version \"1\"...</td>\n",
       "      <td>1623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868407</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>61223957</td>\n",
       "      <td>61228429</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000059326\"; gene_version \"7\"...</td>\n",
       "      <td>4472</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22436 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seq_name          source feature     start       end score strand  \\\n",
       "6              1  ensembl_havana    gene   3205901   3671498     .      -   \n",
       "70             1  ensembl_havana    gene   3999557   4409241     .      -   \n",
       "160            1  ensembl_havana    gene   4490931   4497354     .      -   \n",
       "283            1  ensembl_havana    gene   4773206   4785739     .      -   \n",
       "354            1  ensembl_havana    gene   4807788   4848410     .      +   \n",
       "...          ...             ...     ...       ...       ...   ...    ...   \n",
       "1868189       19  ensembl_havana    gene  60864051  60874556     .      -   \n",
       "1868209       19  ensembl_havana    gene  60889749  61095968     .      +   \n",
       "1868272       19  ensembl_havana    gene  61053840  61140840     .      -   \n",
       "1868395       19  ensembl_havana    gene  61174686  61176309     .      -   \n",
       "1868407       19  ensembl_havana    gene  61223957  61228429     .      -   \n",
       "\n",
       "        frame                                          attribute  gene_length  \n",
       "6           .  gene_id \"ENSMUSG00000051951\"; gene_version \"5\"...       465597  \n",
       "70          .  gene_id \"ENSMUSG00000025900\"; gene_version \"13...       409684  \n",
       "160         .  gene_id \"ENSMUSG00000025902\"; gene_version \"13...         6423  \n",
       "283         .  gene_id \"ENSMUSG00000033845\"; gene_version \"13...        12533  \n",
       "354         .  gene_id \"ENSMUSG00000025903\"; gene_version \"14...        40622  \n",
       "...       ...                                                ...          ...  \n",
       "1868189     .  gene_id \"ENSMUSG00000024997\"; gene_version \"7\"...        10505  \n",
       "1868209     .  gene_id \"ENSMUSG00000003228\"; gene_version \"9\"...       206219  \n",
       "1868272     .  gene_id \"ENSMUSG00000074733\"; gene_version \"14...        87000  \n",
       "1868395     .  gene_id \"ENSMUSG00000094649\"; gene_version \"1\"...         1623  \n",
       "1868407     .  gene_id \"ENSMUSG00000059326\"; gene_version \"7\"...         4472  \n",
       "\n",
       "[22436 rows x 10 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "start          7.493065e+07\n",
       "end            7.497784e+07\n",
       "gene_length    4.718941e+04\n",
       "dtype: float64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "start          73367660.5\n",
       "end            73407314.5\n",
       "gene_length       16308.5\n",
       "dtype: float64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "16308.5"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "47189.4075147085"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "count    2.243600e+04\n",
       "mean     4.718941e+04\n",
       "std      1.095276e+05\n",
       "min      9.000000e+00\n",
       "25%      5.905750e+03\n",
       "50%      1.630850e+04\n",
       "75%      4.379100e+04\n",
       "max      2.960899e+06\n",
       "Name: gene_length, dtype: float64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_name</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>881840</th>\n",
       "      <td>7</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>130162451</td>\n",
       "      <td>133123350</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000030849\"; gene_version \"18...</td>\n",
       "      <td>2960899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>452202</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>75941238</td>\n",
       "      <td>78211961</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000028399\"; gene_version \"18...</td>\n",
       "      <td>2270723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>301809</th>\n",
       "      <td>X</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>82948870</td>\n",
       "      <td>85206141</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000045103\"; gene_version \"19...</td>\n",
       "      <td>2257271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>673541</th>\n",
       "      <td>6</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>45059357</td>\n",
       "      <td>47304213</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000039419\"; gene_version \"17...</td>\n",
       "      <td>2244856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1626801</th>\n",
       "      <td>16</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>39984361</td>\n",
       "      <td>42181679</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000061080\"; gene_version \"11...</td>\n",
       "      <td>2197318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158839</th>\n",
       "      <td>2</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>40595246</td>\n",
       "      <td>42653624</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000049252\"; gene_version \"17...</td>\n",
       "      <td>2058378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232934</th>\n",
       "      <td>2</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>140395309</td>\n",
       "      <td>142392966</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000068205\"; gene_version \"14...</td>\n",
       "      <td>1997657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>836474</th>\n",
       "      <td>7</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>90476672</td>\n",
       "      <td>92449247</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000052572\"; gene_version \"19...</td>\n",
       "      <td>1972575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999774</th>\n",
       "      <td>8</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>15892537</td>\n",
       "      <td>17535586</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000060924\"; gene_version \"15...</td>\n",
       "      <td>1643049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1498920</th>\n",
       "      <td>12</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>88722876</td>\n",
       "      <td>90334935</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000066392\"; gene_version \"12...</td>\n",
       "      <td>1612059</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        seq_name          source feature      start        end score strand  \\\n",
       "881840         7  ensembl_havana    gene  130162451  133123350     .      -   \n",
       "452202         4  ensembl_havana    gene   75941238   78211961     .      -   \n",
       "301809         X  ensembl_havana    gene   82948870   85206141     .      +   \n",
       "673541         6  ensembl_havana    gene   45059357   47304213     .      +   \n",
       "1626801       16  ensembl_havana    gene   39984361   42181679     .      +   \n",
       "158839         2  ensembl_havana    gene   40595246   42653624     .      -   \n",
       "232934         2  ensembl_havana    gene  140395309  142392966     .      +   \n",
       "836474         7  ensembl_havana    gene   90476672   92449247     .      +   \n",
       "999774         8  ensembl_havana    gene   15892537   17535586     .      -   \n",
       "1498920       12  ensembl_havana    gene   88722876   90334935     .      +   \n",
       "\n",
       "        frame                                          attribute  gene_length  \n",
       "881840      .  gene_id \"ENSMUSG00000030849\"; gene_version \"18...      2960899  \n",
       "452202      .  gene_id \"ENSMUSG00000028399\"; gene_version \"18...      2270723  \n",
       "301809      .  gene_id \"ENSMUSG00000045103\"; gene_version \"19...      2257271  \n",
       "673541      .  gene_id \"ENSMUSG00000039419\"; gene_version \"17...      2244856  \n",
       "1626801     .  gene_id \"ENSMUSG00000061080\"; gene_version \"11...      2197318  \n",
       "158839      .  gene_id \"ENSMUSG00000049252\"; gene_version \"17...      2058378  \n",
       "232934      .  gene_id \"ENSMUSG00000068205\"; gene_version \"14...      1997657  \n",
       "836474      .  gene_id \"ENSMUSG00000052572\"; gene_version \"19...      1972575  \n",
       "999774      .  gene_id \"ENSMUSG00000060924\"; gene_version \"15...      1643049  \n",
       "1498920     .  gene_id \"ENSMUSG00000066392\"; gene_version \"12...      1612059  "
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Assign appropriate column names https://www.ensembl.org/info/website/upload/gff.html\n",
    "gtf.columns = ['seq_name', 'source', 'feature','start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "print(gtf) \n",
    "\n",
    "#Find the unique values for 'source' column\n",
    "print(gtf['source'].unique())\n",
    "\n",
    "#Subset all values from 'ensembl_havana' and assign to new df called 'ens_df'\n",
    "ens_df = gtf[gtf.source == 'ensembl_havana']\n",
    "\n",
    "#What is the size of ens_df?\n",
    "ens_df.info\n",
    "ens_df.head()\n",
    "\n",
    "#Group ens_df by 'feature' column & apply the 'count' function to get the number of rows in each feature\n",
    "ens_df.groupby('feature').count()\n",
    "\n",
    "#Create a new df called 'genes' by subsetting all rows with 'gene' as a feature\n",
    "gene_df= ens_df[ens_df.feature == 'gene'].copy()\n",
    "display(gene_df)\n",
    "\n",
    "#Create a new column called 'gene_length' and calculate the length of each gene\n",
    "gene_df['gene_length']= gene_df['end']- gene_df['start']\n",
    "display(gene_df)\n",
    "\n",
    "#Calculate the mean and median of gene lengths\n",
    "display(gene_df.mean(axis=0))\n",
    "display(gene_df.median(axis=0))\n",
    "display(np.median(gene_df['gene_length']))\n",
    "display(np.mean(gene_df['gene_length']))\n",
    "display(gene_df['gene_length'].describe())\n",
    "\n",
    "\n",
    "#Find the top 10 longest genes\n",
    "gene_df.sort_values(by=['gene_length'], ascending=False).head(10)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "stainless-netscape",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>gene_id \"ENSMUSG00000051951\"</td>\n",
       "      <td>gene_version \"5\"</td>\n",
       "      <td>gene_name \"Xkr4\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td>havana_gene \"OTTMUSG00000026353\"</td>\n",
       "      <td>havana_gene_version \"2\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>gene_id \"ENSMUSG00000025900\"</td>\n",
       "      <td>gene_version \"13\"</td>\n",
       "      <td>gene_name \"Rp1\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td>havana_gene \"OTTMUSG00000049985\"</td>\n",
       "      <td>havana_gene_version \"3\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>gene_id \"ENSMUSG00000025902\"</td>\n",
       "      <td>gene_version \"13\"</td>\n",
       "      <td>gene_name \"Sox17\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td>havana_gene \"OTTMUSG00000050014\"</td>\n",
       "      <td>havana_gene_version \"7\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>gene_id \"ENSMUSG00000033845\"</td>\n",
       "      <td>gene_version \"13\"</td>\n",
       "      <td>gene_name \"Mrpl15\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td>havana_gene \"OTTMUSG00000029329\"</td>\n",
       "      <td>havana_gene_version \"3\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>gene_id \"ENSMUSG00000025903\"</td>\n",
       "      <td>gene_version \"14\"</td>\n",
       "      <td>gene_name \"Lypla1\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td>havana_gene \"OTTMUSG00000021562\"</td>\n",
       "      <td>havana_gene_version \"4\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868189</th>\n",
       "      <td>gene_id \"ENSMUSG00000024997\"</td>\n",
       "      <td>gene_version \"7\"</td>\n",
       "      <td>gene_name \"Prdx3\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td>havana_gene \"OTTMUSG00000021212\"</td>\n",
       "      <td>havana_gene_version \"1\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868209</th>\n",
       "      <td>gene_id \"ENSMUSG00000003228\"</td>\n",
       "      <td>gene_version \"9\"</td>\n",
       "      <td>gene_name \"Grk5\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td>havana_gene \"OTTMUSG00000074399\"</td>\n",
       "      <td>havana_gene_version \"1\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868272</th>\n",
       "      <td>gene_id \"ENSMUSG00000074733\"</td>\n",
       "      <td>gene_version \"14\"</td>\n",
       "      <td>gene_name \"Zfp950\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td>havana_gene \"OTTMUSG00000033031\"</td>\n",
       "      <td>havana_gene_version \"5\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868395</th>\n",
       "      <td>gene_id \"ENSMUSG00000094649\"</td>\n",
       "      <td>gene_version \"1\"</td>\n",
       "      <td>gene_name \"Gm7102\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td>havana_gene \"OTTMUSG00000042892\"</td>\n",
       "      <td>havana_gene_version \"1\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868407</th>\n",
       "      <td>gene_id \"ENSMUSG00000059326\"</td>\n",
       "      <td>gene_version \"7\"</td>\n",
       "      <td>gene_name \"Csf2ra\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td>havana_gene \"OTTMUSG00000074402\"</td>\n",
       "      <td>havana_gene_version \"1\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22436 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    0                   1  \\\n",
       "6        gene_id \"ENSMUSG00000051951\"    gene_version \"5\"   \n",
       "70       gene_id \"ENSMUSG00000025900\"   gene_version \"13\"   \n",
       "160      gene_id \"ENSMUSG00000025902\"   gene_version \"13\"   \n",
       "283      gene_id \"ENSMUSG00000033845\"   gene_version \"13\"   \n",
       "354      gene_id \"ENSMUSG00000025903\"   gene_version \"14\"   \n",
       "...                               ...                 ...   \n",
       "1868189  gene_id \"ENSMUSG00000024997\"    gene_version \"7\"   \n",
       "1868209  gene_id \"ENSMUSG00000003228\"    gene_version \"9\"   \n",
       "1868272  gene_id \"ENSMUSG00000074733\"   gene_version \"14\"   \n",
       "1868395  gene_id \"ENSMUSG00000094649\"    gene_version \"1\"   \n",
       "1868407  gene_id \"ENSMUSG00000059326\"    gene_version \"7\"   \n",
       "\n",
       "                           2                              3  \\\n",
       "6           gene_name \"Xkr4\"   gene_source \"ensembl_havana\"   \n",
       "70           gene_name \"Rp1\"   gene_source \"ensembl_havana\"   \n",
       "160        gene_name \"Sox17\"   gene_source \"ensembl_havana\"   \n",
       "283       gene_name \"Mrpl15\"   gene_source \"ensembl_havana\"   \n",
       "354       gene_name \"Lypla1\"   gene_source \"ensembl_havana\"   \n",
       "...                      ...                            ...   \n",
       "1868189    gene_name \"Prdx3\"   gene_source \"ensembl_havana\"   \n",
       "1868209     gene_name \"Grk5\"   gene_source \"ensembl_havana\"   \n",
       "1868272   gene_name \"Zfp950\"   gene_source \"ensembl_havana\"   \n",
       "1868395   gene_name \"Gm7102\"   gene_source \"ensembl_havana\"   \n",
       "1868407   gene_name \"Csf2ra\"   gene_source \"ensembl_havana\"   \n",
       "\n",
       "                                      4                                  5  \\\n",
       "6         gene_biotype \"protein_coding\"   havana_gene \"OTTMUSG00000026353\"   \n",
       "70        gene_biotype \"protein_coding\"   havana_gene \"OTTMUSG00000049985\"   \n",
       "160       gene_biotype \"protein_coding\"   havana_gene \"OTTMUSG00000050014\"   \n",
       "283       gene_biotype \"protein_coding\"   havana_gene \"OTTMUSG00000029329\"   \n",
       "354       gene_biotype \"protein_coding\"   havana_gene \"OTTMUSG00000021562\"   \n",
       "...                                 ...                                ...   \n",
       "1868189   gene_biotype \"protein_coding\"   havana_gene \"OTTMUSG00000021212\"   \n",
       "1868209   gene_biotype \"protein_coding\"   havana_gene \"OTTMUSG00000074399\"   \n",
       "1868272   gene_biotype \"protein_coding\"   havana_gene \"OTTMUSG00000033031\"   \n",
       "1868395   gene_biotype \"protein_coding\"   havana_gene \"OTTMUSG00000042892\"   \n",
       "1868407   gene_biotype \"protein_coding\"   havana_gene \"OTTMUSG00000074402\"   \n",
       "\n",
       "                                6 7   \n",
       "6         havana_gene_version \"2\"     \n",
       "70        havana_gene_version \"3\"     \n",
       "160       havana_gene_version \"7\"     \n",
       "283       havana_gene_version \"3\"     \n",
       "354       havana_gene_version \"4\"     \n",
       "...                           ... ..  \n",
       "1868189   havana_gene_version \"1\"     \n",
       "1868209   havana_gene_version \"1\"     \n",
       "1868272   havana_gene_version \"5\"     \n",
       "1868395   havana_gene_version \"1\"     \n",
       "1868407   havana_gene_version \"1\"     \n",
       "\n",
       "[22436 rows x 8 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_name</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3205901</td>\n",
       "      <td>3671498</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000051951\"; gene_version \"5\"...</td>\n",
       "      <td>465597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3999557</td>\n",
       "      <td>4409241</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000025900\"; gene_version \"13...</td>\n",
       "      <td>409684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>4490931</td>\n",
       "      <td>4497354</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000025902\"; gene_version \"13...</td>\n",
       "      <td>6423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>4773206</td>\n",
       "      <td>4785739</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000033845\"; gene_version \"13...</td>\n",
       "      <td>12533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>4807788</td>\n",
       "      <td>4848410</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000025903\"; gene_version \"14...</td>\n",
       "      <td>40622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868189</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>60864051</td>\n",
       "      <td>60874556</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000024997\"; gene_version \"7\"...</td>\n",
       "      <td>10505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868209</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>60889749</td>\n",
       "      <td>61095968</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000003228\"; gene_version \"9\"...</td>\n",
       "      <td>206219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868272</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>61053840</td>\n",
       "      <td>61140840</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000074733\"; gene_version \"14...</td>\n",
       "      <td>87000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868395</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>61174686</td>\n",
       "      <td>61176309</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000094649\"; gene_version \"1\"...</td>\n",
       "      <td>1623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868407</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>61223957</td>\n",
       "      <td>61228429</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000059326\"; gene_version \"7\"...</td>\n",
       "      <td>4472</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22436 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seq_name          source feature     start       end score strand  \\\n",
       "6              1  ensembl_havana    gene   3205901   3671498     .      -   \n",
       "70             1  ensembl_havana    gene   3999557   4409241     .      -   \n",
       "160            1  ensembl_havana    gene   4490931   4497354     .      -   \n",
       "283            1  ensembl_havana    gene   4773206   4785739     .      -   \n",
       "354            1  ensembl_havana    gene   4807788   4848410     .      +   \n",
       "...          ...             ...     ...       ...       ...   ...    ...   \n",
       "1868189       19  ensembl_havana    gene  60864051  60874556     .      -   \n",
       "1868209       19  ensembl_havana    gene  60889749  61095968     .      +   \n",
       "1868272       19  ensembl_havana    gene  61053840  61140840     .      -   \n",
       "1868395       19  ensembl_havana    gene  61174686  61176309     .      -   \n",
       "1868407       19  ensembl_havana    gene  61223957  61228429     .      -   \n",
       "\n",
       "        frame                                          attribute  gene_length  \n",
       "6           .  gene_id \"ENSMUSG00000051951\"; gene_version \"5\"...       465597  \n",
       "70          .  gene_id \"ENSMUSG00000025900\"; gene_version \"13...       409684  \n",
       "160         .  gene_id \"ENSMUSG00000025902\"; gene_version \"13...         6423  \n",
       "283         .  gene_id \"ENSMUSG00000033845\"; gene_version \"13...        12533  \n",
       "354         .  gene_id \"ENSMUSG00000025903\"; gene_version \"14...        40622  \n",
       "...       ...                                                ...          ...  \n",
       "1868189     .  gene_id \"ENSMUSG00000024997\"; gene_version \"7\"...        10505  \n",
       "1868209     .  gene_id \"ENSMUSG00000003228\"; gene_version \"9\"...       206219  \n",
       "1868272     .  gene_id \"ENSMUSG00000074733\"; gene_version \"14...        87000  \n",
       "1868395     .  gene_id \"ENSMUSG00000094649\"; gene_version \"1\"...         1623  \n",
       "1868407     .  gene_id \"ENSMUSG00000059326\"; gene_version \"7\"...         4472  \n",
       "\n",
       "[22436 rows x 10 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_name</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_length</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3205901</td>\n",
       "      <td>3671498</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000051951\"; gene_version \"5\"...</td>\n",
       "      <td>465597</td>\n",
       "      <td>gene_id \"ENSMUSG00000051951\"</td>\n",
       "      <td>gene_version \"5\"</td>\n",
       "      <td>gene_name \"Xkr4\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td>havana_gene \"OTTMUSG00000026353\"</td>\n",
       "      <td>havana_gene_version \"2\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3999557</td>\n",
       "      <td>4409241</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000025900\"; gene_version \"13...</td>\n",
       "      <td>409684</td>\n",
       "      <td>gene_id \"ENSMUSG00000025900\"</td>\n",
       "      <td>gene_version \"13\"</td>\n",
       "      <td>gene_name \"Rp1\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td>havana_gene \"OTTMUSG00000049985\"</td>\n",
       "      <td>havana_gene_version \"3\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>4490931</td>\n",
       "      <td>4497354</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000025902\"; gene_version \"13...</td>\n",
       "      <td>6423</td>\n",
       "      <td>gene_id \"ENSMUSG00000025902\"</td>\n",
       "      <td>gene_version \"13\"</td>\n",
       "      <td>gene_name \"Sox17\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td>havana_gene \"OTTMUSG00000050014\"</td>\n",
       "      <td>havana_gene_version \"7\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>4773206</td>\n",
       "      <td>4785739</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000033845\"; gene_version \"13...</td>\n",
       "      <td>12533</td>\n",
       "      <td>gene_id \"ENSMUSG00000033845\"</td>\n",
       "      <td>gene_version \"13\"</td>\n",
       "      <td>gene_name \"Mrpl15\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td>havana_gene \"OTTMUSG00000029329\"</td>\n",
       "      <td>havana_gene_version \"3\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>4807788</td>\n",
       "      <td>4848410</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000025903\"; gene_version \"14...</td>\n",
       "      <td>40622</td>\n",
       "      <td>gene_id \"ENSMUSG00000025903\"</td>\n",
       "      <td>gene_version \"14\"</td>\n",
       "      <td>gene_name \"Lypla1\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td>havana_gene \"OTTMUSG00000021562\"</td>\n",
       "      <td>havana_gene_version \"4\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868189</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>60864051</td>\n",
       "      <td>60874556</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000024997\"; gene_version \"7\"...</td>\n",
       "      <td>10505</td>\n",
       "      <td>gene_id \"ENSMUSG00000024997\"</td>\n",
       "      <td>gene_version \"7\"</td>\n",
       "      <td>gene_name \"Prdx3\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td>havana_gene \"OTTMUSG00000021212\"</td>\n",
       "      <td>havana_gene_version \"1\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868209</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>60889749</td>\n",
       "      <td>61095968</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000003228\"; gene_version \"9\"...</td>\n",
       "      <td>206219</td>\n",
       "      <td>gene_id \"ENSMUSG00000003228\"</td>\n",
       "      <td>gene_version \"9\"</td>\n",
       "      <td>gene_name \"Grk5\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td>havana_gene \"OTTMUSG00000074399\"</td>\n",
       "      <td>havana_gene_version \"1\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868272</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>61053840</td>\n",
       "      <td>61140840</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000074733\"; gene_version \"14...</td>\n",
       "      <td>87000</td>\n",
       "      <td>gene_id \"ENSMUSG00000074733\"</td>\n",
       "      <td>gene_version \"14\"</td>\n",
       "      <td>gene_name \"Zfp950\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td>havana_gene \"OTTMUSG00000033031\"</td>\n",
       "      <td>havana_gene_version \"5\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868395</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>61174686</td>\n",
       "      <td>61176309</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000094649\"; gene_version \"1\"...</td>\n",
       "      <td>1623</td>\n",
       "      <td>gene_id \"ENSMUSG00000094649\"</td>\n",
       "      <td>gene_version \"1\"</td>\n",
       "      <td>gene_name \"Gm7102\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td>havana_gene \"OTTMUSG00000042892\"</td>\n",
       "      <td>havana_gene_version \"1\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868407</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>61223957</td>\n",
       "      <td>61228429</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000059326\"; gene_version \"7\"...</td>\n",
       "      <td>4472</td>\n",
       "      <td>gene_id \"ENSMUSG00000059326\"</td>\n",
       "      <td>gene_version \"7\"</td>\n",
       "      <td>gene_name \"Csf2ra\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td>havana_gene \"OTTMUSG00000074402\"</td>\n",
       "      <td>havana_gene_version \"1\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22436 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seq_name          source feature     start       end score strand  \\\n",
       "6              1  ensembl_havana    gene   3205901   3671498     .      -   \n",
       "70             1  ensembl_havana    gene   3999557   4409241     .      -   \n",
       "160            1  ensembl_havana    gene   4490931   4497354     .      -   \n",
       "283            1  ensembl_havana    gene   4773206   4785739     .      -   \n",
       "354            1  ensembl_havana    gene   4807788   4848410     .      +   \n",
       "...          ...             ...     ...       ...       ...   ...    ...   \n",
       "1868189       19  ensembl_havana    gene  60864051  60874556     .      -   \n",
       "1868209       19  ensembl_havana    gene  60889749  61095968     .      +   \n",
       "1868272       19  ensembl_havana    gene  61053840  61140840     .      -   \n",
       "1868395       19  ensembl_havana    gene  61174686  61176309     .      -   \n",
       "1868407       19  ensembl_havana    gene  61223957  61228429     .      -   \n",
       "\n",
       "        frame                                          attribute  gene_length  \\\n",
       "6           .  gene_id \"ENSMUSG00000051951\"; gene_version \"5\"...       465597   \n",
       "70          .  gene_id \"ENSMUSG00000025900\"; gene_version \"13...       409684   \n",
       "160         .  gene_id \"ENSMUSG00000025902\"; gene_version \"13...         6423   \n",
       "283         .  gene_id \"ENSMUSG00000033845\"; gene_version \"13...        12533   \n",
       "354         .  gene_id \"ENSMUSG00000025903\"; gene_version \"14...        40622   \n",
       "...       ...                                                ...          ...   \n",
       "1868189     .  gene_id \"ENSMUSG00000024997\"; gene_version \"7\"...        10505   \n",
       "1868209     .  gene_id \"ENSMUSG00000003228\"; gene_version \"9\"...       206219   \n",
       "1868272     .  gene_id \"ENSMUSG00000074733\"; gene_version \"14...        87000   \n",
       "1868395     .  gene_id \"ENSMUSG00000094649\"; gene_version \"1\"...         1623   \n",
       "1868407     .  gene_id \"ENSMUSG00000059326\"; gene_version \"7\"...         4472   \n",
       "\n",
       "                                    0                   1  \\\n",
       "6        gene_id \"ENSMUSG00000051951\"    gene_version \"5\"   \n",
       "70       gene_id \"ENSMUSG00000025900\"   gene_version \"13\"   \n",
       "160      gene_id \"ENSMUSG00000025902\"   gene_version \"13\"   \n",
       "283      gene_id \"ENSMUSG00000033845\"   gene_version \"13\"   \n",
       "354      gene_id \"ENSMUSG00000025903\"   gene_version \"14\"   \n",
       "...                               ...                 ...   \n",
       "1868189  gene_id \"ENSMUSG00000024997\"    gene_version \"7\"   \n",
       "1868209  gene_id \"ENSMUSG00000003228\"    gene_version \"9\"   \n",
       "1868272  gene_id \"ENSMUSG00000074733\"   gene_version \"14\"   \n",
       "1868395  gene_id \"ENSMUSG00000094649\"    gene_version \"1\"   \n",
       "1868407  gene_id \"ENSMUSG00000059326\"    gene_version \"7\"   \n",
       "\n",
       "                           2                              3  \\\n",
       "6           gene_name \"Xkr4\"   gene_source \"ensembl_havana\"   \n",
       "70           gene_name \"Rp1\"   gene_source \"ensembl_havana\"   \n",
       "160        gene_name \"Sox17\"   gene_source \"ensembl_havana\"   \n",
       "283       gene_name \"Mrpl15\"   gene_source \"ensembl_havana\"   \n",
       "354       gene_name \"Lypla1\"   gene_source \"ensembl_havana\"   \n",
       "...                      ...                            ...   \n",
       "1868189    gene_name \"Prdx3\"   gene_source \"ensembl_havana\"   \n",
       "1868209     gene_name \"Grk5\"   gene_source \"ensembl_havana\"   \n",
       "1868272   gene_name \"Zfp950\"   gene_source \"ensembl_havana\"   \n",
       "1868395   gene_name \"Gm7102\"   gene_source \"ensembl_havana\"   \n",
       "1868407   gene_name \"Csf2ra\"   gene_source \"ensembl_havana\"   \n",
       "\n",
       "                                      4                                  5  \\\n",
       "6         gene_biotype \"protein_coding\"   havana_gene \"OTTMUSG00000026353\"   \n",
       "70        gene_biotype \"protein_coding\"   havana_gene \"OTTMUSG00000049985\"   \n",
       "160       gene_biotype \"protein_coding\"   havana_gene \"OTTMUSG00000050014\"   \n",
       "283       gene_biotype \"protein_coding\"   havana_gene \"OTTMUSG00000029329\"   \n",
       "354       gene_biotype \"protein_coding\"   havana_gene \"OTTMUSG00000021562\"   \n",
       "...                                 ...                                ...   \n",
       "1868189   gene_biotype \"protein_coding\"   havana_gene \"OTTMUSG00000021212\"   \n",
       "1868209   gene_biotype \"protein_coding\"   havana_gene \"OTTMUSG00000074399\"   \n",
       "1868272   gene_biotype \"protein_coding\"   havana_gene \"OTTMUSG00000033031\"   \n",
       "1868395   gene_biotype \"protein_coding\"   havana_gene \"OTTMUSG00000042892\"   \n",
       "1868407   gene_biotype \"protein_coding\"   havana_gene \"OTTMUSG00000074402\"   \n",
       "\n",
       "                                6 7   \n",
       "6         havana_gene_version \"2\"     \n",
       "70        havana_gene_version \"3\"     \n",
       "160       havana_gene_version \"7\"     \n",
       "283       havana_gene_version \"3\"     \n",
       "354       havana_gene_version \"4\"     \n",
       "...                           ... ..  \n",
       "1868189   havana_gene_version \"1\"     \n",
       "1868209   havana_gene_version \"1\"     \n",
       "1868272   havana_gene_version \"5\"     \n",
       "1868395   havana_gene_version \"1\"     \n",
       "1868407   havana_gene_version \"1\"     \n",
       "\n",
       "[22436 rows x 18 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Find the lowest 20 shortest genes\n",
    "gene_df.sort_values(by=['gene_length'], ascending=True).head(20)\n",
    "\n",
    "#Split attributes column on ';' and create new dataframe, merge this onto the ens_gtf dataframe using pandas merge \n",
    "attributes_df=gene_df['attribute'].str.split(';', expand=True)\n",
    "display(attributes_df)\n",
    "display(gene_df)\n",
    "\n",
    "genes_full_attributes = pd.concat([gene_df, attributes_df], axis=1)\n",
    "\n",
    "display(genes_full_attributes)\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "colonial-pantyhose",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Count the number of genes per chromosome, and add the number of genes per chomosome onto the idxdf from earlier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "behavioral-simple",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
